File: python/src/openmc_metal/simulation.py
"""Main k-eigenvalue simulation driver."""

import time
import struct as struct_mod

from .types import PARTICLE_SIZE, pack_simulation_params
from .engine import MetalEngine
from .geometry import GeometryData, source_sampler
from .particles import ParticleBuffer
from .transport import TransportKernels
from .collision import FissionBank, CollisionKernel
from .tally import TallyManager
from .statistics import keff_statistics, compute_stats
from .keigenvalue import shannon_entropy, is_converged


class SimulationResult:
    """Aggregated results from a completed k-eigenvalue simulation."""

    def __init__(self, k_eff, flux_stats, fission_stats, wall_time,
                 particles_per_second, total_particles, num_active_batches,
                 gpu_name, batch_keff_history, entropy_history):
        self.k_eff = k_eff  # (mean, std_dev, (ci95_lo, ci95_hi))
        self.flux_stats = flux_stats
        self.fission_stats = fission_stats
        self.wall_time = wall_time
        self.particles_per_second = particles_per_second
        self.total_particles = total_particles
        self.num_active_batches = num_active_batches
        self.gpu_name = gpu_name
        self.batch_keff_history = batch_keff_history
        self.entropy_history = entropy_history

    def print_summary(self, reference=None):
        """Print formatted simulation results."""
        print("\n" + "=" * 60)
        print("SIMULATION RESULTS")
        print("=" * 60)
        mean, std_dev, ci95 = self.k_eff
        print(f"  k-eff = {mean:.5f} +/- {std_dev:.5f}")
        print(f"  95% CI: [{ci95[0]:.5f}, {ci95[1]:.5f}]")
        if reference:
            delta_pcm = (mean - reference) * 1e5
            print(f"  Reference: {reference:.5f}  (delta = {delta_pcm:+.1f} pcm)")
        print(f"\n  Performance: {self.particles_per_second:.0f} particles/sec")
        print(f"  Total particles: {self.total_particles}")
        print(f"  Wall time: {self.wall_time:.2f} sec")
        print(f"  Active batches: {self.num_active_batches}")
        print(f"  GPU: {self.gpu_name}")
        print("=" * 60)


class Simulation:
    """k-eigenvalue Monte Carlo simulation using Apple Metal GPU kernels."""

    def __init__(self, engine: MetalEngine, geometry: GeometryData,
                 materials_buffer, num_particles: int = 100_000,
                 num_batches: int = 100, num_inactive: int = 50,
                 num_groups: int = 7):
        self.engine = engine
        self.geometry = geometry
        self.materials_buffer = materials_buffer
        self.num_particles = num_particles
        self.num_batches = num_batches
        self.num_inactive = num_inactive
        self.num_groups = num_groups

        self.kernels = TransportKernels(engine)
        self.collision = CollisionKernel(engine)
        self.tally = TallyManager(engine, geometry.num_cells, num_groups)
        self.fission_bank = FissionBank(engine, num_particles * 3)
        self.particle_buf = ParticleBuffer(engine, num_particles)
        self.params_buffer = engine.make_buffer(32)
        self.lost_count_buffer = engine.make_buffer(4)
        self.current_keff = 1.0
        self._write_params()

    def _write_params(self):
        """Write current simulation params to GPU buffer."""
        data = pack_simulation_params(
            num_particles=self.num_particles,
            num_batches=self.num_batches,
            num_inactive=self.num_inactive,
            num_groups=self.num_groups,
            num_cells=self.geometry.num_cells,
            num_surfaces=self.geometry.num_surfaces,
            num_materials=7,
            k_eff=self.current_keff
        )
        self.engine.write_buffer_bytes(self.params_buffer, data)

    def run(self, source: list, verbose: bool = False) -> SimulationResult:
        """Execute the full k-eigenvalue simulation."""
        wall_start = time.time()
        total_particles = 0
        current_source = source
        active_batches = self.num_batches - self.num_inactive

        print("Starting k-eigenvalue simulation")
        print(f"  Particles per batch: {self.num_particles}")
        print(f"  Inactive batches:    {self.num_inactive}")
        print(f"  Active batches:      {active_batches}")
        print(f"  Energy groups:       {self.num_groups}")
        print(f"  Cells:               {self.geometry.num_cells}")
        print()

        batch_keff = []
        entropy_history = []

        for batch_idx in range(self.num_batches):
            is_active = batch_idx >= self.num_inactive

            # Reset per-batch state
            self.fission_bank.reset()
            self.tally.reset_for_batch()
            self.engine.zero_buffer(self.lost_count_buffer, 4)

            # Seed particles from current source
            self.particle_buf.seed_from_source(current_source, batch_idx)

            # Update params with current k-eff
            self._write_params()

            # Run transport loop
            iterations = self._transport_batch()

            # Read fission bank
            fission_sites = self.fission_bank.sites()
            fission_count = len(fission_sites)

            # Compute batch k-eff: fission bank already includes nu (the
            # collision kernel banks nu*weight sites per fission), so
            # k_eff = fission_sites / source_particles directly.
            batch_k = fission_count / self.num_particles if fission_count > 0 else 0.0

            # Collect tallies for active batches
            if is_active:
                self.tally.collect_batch()
                batch_keff.append(batch_k)

            # Shannon entropy
            if fission_sites:
                positions = [pos for pos, _ in fission_sites]
                entropy = shannon_entropy(positions)
                entropy_history.append(entropy)

            # Update k-eff
            self.current_keff = float(batch_k)

            # Resample fission bank for next batch
            current_source = self._resample_fission_bank(fission_sites)

            total_particles += self.num_particles

            # Clear tally history after inactive phase
            if batch_idx == self.num_inactive - 1:
                self.tally.clear_batch_history()

            # Print progress
            lost_mv = self.engine.buffer_view(self.lost_count_buffer)
            lost_count = struct_mod.unpack_from('<I', lost_mv, 0)[0]
            if verbose or batch_idx % 10 == 0 or batch_idx == self.num_batches - 1:
                status = "Active  " if is_active else "Inactive"
                converged = " [converged]" if is_converged(entropy_history) else ""
                print(f"Batch {batch_idx+1:3d}/{self.num_batches} [{status}] "
                      f"k-eff = {batch_k:.5f}  fission = {fission_count}  "
                      f"lost = {lost_count}  iter = {iterations}{converged}")

        wall_time = time.time() - wall_start
        particles_per_sec = total_particles / wall_time

        # Compute statistics
        k_eff_stats = keff_statistics(batch_keff)
        flux_stats = compute_stats(self.tally.batch_flux)
        fission_stats = compute_stats(self.tally.batch_fission)

        return SimulationResult(
            k_eff=k_eff_stats,
            flux_stats=flux_stats,
            fission_stats=fission_stats,
            wall_time=wall_time,
            particles_per_second=particles_per_sec,
            total_particles=total_particles,
            num_active_batches=active_batches,
            gpu_name=self.engine.gpu_name,
            batch_keff_history=batch_keff,
            entropy_history=entropy_history,
        )

    def _transport_batch(self) -> int:
        """Run event-based transport loop until all particles dead.

        Uses 3 fused GPU kernels per step (down from 5):
          1. xs_lookup_and_distance (fused XS lookup + distance sampling)
          2. move_particle (standalone - complex geometry logic)
          3. collision_and_tally (fused collision + tally scoring)

        Metal serial queues guarantee in-order execution, so we only
        sync (waitUntilCompleted) when we need to read results on CPU.
        Between syncs, command buffers are committed without waiting,
        allowing the GPU to pipeline work continuously.
        """
        max_steps = 5000
        check_interval = 25  # check all_dead every N steps

        last_cmd = None

        for step in range(max_steps):
            cmd = self.engine.command_queue.commandBuffer()

            # 1. Fused XS Lookup + Distance to Collision
            self.kernels.dispatch_xs_lookup_and_distance(
                self.particle_buf.buffer, self.materials_buffer,
                self.geometry.cell_buffer, self.params_buffer,
                self.num_particles, cmd
            )

            # 2. Move Particle (standalone - complex geometry)
            self.kernels.dispatch_move(
                self.particle_buf.buffer,
                self.geometry.surface_buffer,
                self.geometry.cell_buffer,
                self.geometry.cell_surface_buffer,
                self.params_buffer, self.lost_count_buffer,
                self.num_particles, cmd
            )

            # 3. Fused Collision + Tally
            self.collision.dispatch_fused(
                self.particle_buf.buffer, self.materials_buffer,
                self.fission_bank, self.params_buffer,
                self.tally.flux_buffer, self.tally.fission_buffer,
                self.num_particles, cmd
            )

            cmd.commit()
            last_cmd = cmd

            # Only sync and check periodically to reduce CPU-GPU overhead
            if (step + 1) % check_interval == 0:
                cmd.waitUntilCompleted()
                if self.particle_buf.all_dead():
                    return step + 1

        # Final sync before returning to ensure all GPU work is complete
        if last_cmd is not None:
            last_cmd.waitUntilCompleted()

        return max_steps

    def _resample_fission_bank(self, sites):
        """Resample fission sites for next batch source via random sampling with replacement."""
        import random
        if not sites:
            return source_sampler(self.num_particles, self.num_groups)

        return random.choices(sites, k=self.num_particles)

File: Sources/OpenMCMetal/Core/Simulation.swift
// Simulation.swift
// OpenMC-Metal
//
// Main simulation driver that orchestrates the event-based GPU transport loop
// for k-eigenvalue calculations. Ties together all kernels:
//
//   xs_lookup -> distance_to_collision -> move_particle -> collision -> tally_score
//       ^                                                                   |
//       |___________________________________________________________________|
//       (alive particles loop back; dead particles exit)
//
// Batch loop:
//   1. Inactive batches: converge fission source, discard tallies
//   2. Active batches:   accumulate tallies, track k-eff per batch
//   3. Post-processing:  compute statistics on k-eff and tallies

import Metal
import Foundation
import simd

// MARK: - SimulationResult

/// Aggregated results from a completed k-eigenvalue simulation.
struct SimulationResult {
    let kEff: (mean: Double, stdDev: Double, ci95: (Double, Double))
    let fluxStats: [(mean: Double, stdDev: Double, relError: Double)]
    let fissionStats: [(mean: Double, stdDev: Double, relError: Double)]
    let wallTime: Double
    let particlesPerSecond: Double
    let totalParticles: UInt64
    let numActiveBatches: Int
    let gpuName: String

    func printSummary(reference: Double? = nil) {
        print("\n" + String(repeating: "=", count: 60))
        print("SIMULATION RESULTS")
        print(String(repeating: "=", count: 60))

        TallyOutput.printKEffResult(stats: kEff, reference: reference,
                                     numBatches: numActiveBatches, wallTime: wallTime)

        print(String(format: "\nPerformance: %.2f particles/sec", particlesPerSecond))
        print(String(format: "Total particles: %llu", totalParticles))
        print("GPU: \(gpuName)")

        let numGroups = 7
        if !fluxStats.isEmpty {
            let numCells = fluxStats.count / numGroups
            TallyOutput.printFluxSummary(stats: fluxStats, numCells: numCells,
                                          numGroups: numGroups)
        }

        if !fissionStats.isEmpty {
            let numCells = fissionStats.count / numGroups
            TallyOutput.printFissionSummary(stats: fissionStats, numCells: numCells,
                                             numGroups: numGroups)
        }
    }
}

// MARK: - SimulationError

enum SimulationError: Error, CustomStringConvertible {
    case pipelineInitFailed(String)
    case tallyInitFailed(String)

    var description: String {
        switch self {
        case .pipelineInitFailed(let msg): return "Pipeline init failed: \(msg)"
        case .tallyInitFailed(let msg):    return "Tally init failed: \(msg)"
        }
    }
}

// MARK: - Simulation

/// k-eigenvalue Monte Carlo simulation using Apple Metal GPU kernels.
///
/// Runs the power-iteration loop:
///   inactive batches -> discard tallies -> active batches -> collect statistics
///
/// Usage:
/// ```swift
/// let sim = try Simulation(engine: engine, geometry: geometry,
///                          materialsBuffer: matBuf,
///                          numParticles: 100_000, numBatches: 100, numInactive: 50)
/// let source = C5G7Geometry.sourceSampler(numParticles: 100_000)
/// let result = sim.run(source: source)
/// result.printSummary(reference: C5G7Geometry.referenceKEff)
/// ```
final class Simulation {
    let engine: MetalEngine
    let geometry: GeometryData
    let materialsBuffer: MTLBuffer
    let numParticles: Int
    let numBatches: Int      // total batches (inactive + active)
    let numInactive: Int     // inactive batches for source convergence
    let numGroups: Int

    // GPU subsystems
    private let kernels: TransportKernels
    private let tally: TallyManager
    private let fissionBank: FissionBank
    private let collisionPipeline: MTLComputePipelineState

    // Persistent GPU buffers
    private let particleBuffer: MTLBuffer
    private let paramsBuffer: MTLBuffer
    private let lostCountBuffer: MTLBuffer

    // Current simulation parameters (mutable for k-eff updates)
    private var params: SimulationParams

    // Per-batch k-eff estimates (active batches only)
    private(set) var batchKEff: [Double] = []

    // Shannon entropy history for convergence diagnostics
    private(set) var shannonEntropy: [Double] = []

    // MARK: Init

    /// Initialize the simulation with geometry, materials, and run parameters.
    ///
    /// - Parameters:
    ///   - engine:       Metal engine for GPU access.
    ///   - geometry:     Packed geometry buffers (surfaces, cells, cell-surfaces).
    ///   - materialsBuffer: Flat float buffer of C5G7 cross-section data.
    ///   - numParticles: Number of particles per batch.
    ///   - numBatches:   Total number of batches (inactive + active).
    ///   - numInactive:  Number of inactive (source convergence) batches.
    ///   - numGroups:    Number of energy groups (default 7 for C5G7).
    init(engine: MetalEngine,
         geometry: GeometryData,
         materialsBuffer: MTLBuffer,
         numParticles: UInt32 = 100_000,
         numBatches: UInt32 = 100,
         numInactive: UInt32 = 50,
         numGroups: UInt32 = 7) throws {

        self.engine          = engine
        self.geometry        = geometry
        self.materialsBuffer = materialsBuffer
        self.numParticles    = Int(numParticles)
        self.numBatches      = Int(numBatches)
        self.numInactive     = Int(numInactive)
        self.numGroups       = Int(numGroups)

        // Transport kernels (xs_lookup, distance_to_collision, move_particle)
        self.kernels = try TransportKernels(engine: engine)

        // Collision kernel pipeline (dispatched manually due to fission bank bindings)
        self.collisionPipeline = try engine.makePipeline(functionName: "collision")

        // Tally manager
        self.tally = try TallyManager(engine: engine,
                                       numCells: Int(geometry.numCells),
                                       numGroups: Int(numGroups))

        // Fission bank -- allow up to 3x particles to avoid overflow
        self.fissionBank = FissionBank(engine: engine,
                                        maxSites: Int(numParticles) * 3)

        // Particle buffer
        self.particleBuffer = engine.makeBuffer(
            length: Int(numParticles) * MemoryLayout<Particle>.stride)
        self.particleBuffer.label = "SimulationParticles"

        // SimulationParams
        self.params = SimulationParams(
            numParticles: numParticles,
            numBatches:   numBatches,
            numInactive:  numInactive,
            numGroups:    numGroups,
            numCells:     geometry.numCells,
            numSurfaces:  geometry.numSurfaces,
            numMaterials: 7,
            kEff:         1.0
        )
        self.paramsBuffer = engine.makeBuffer(length: MemoryLayout<SimulationParams>.stride)
        self.paramsBuffer.label = "SimulationParams"

        // Lost-particle counter
        self.lostCountBuffer = engine.makeBuffer(length: MemoryLayout<UInt32>.stride)
        self.lostCountBuffer.label = "LostParticleCount"

        writeParams()
    }

    // MARK: Run

    /// Execute the full k-eigenvalue simulation.
    ///
    /// - Parameters:
    ///   - source:  Initial source particles as (position, group) pairs.
    ///   - verbose: Print progress for every batch (not just every 10th).
    /// - Returns: Aggregated simulation results with statistics.
    func run(source: [(position: SIMD3<Float>, group: UInt32)],
             verbose: Bool = false) -> SimulationResult {

        let wallStart = CFAbsoluteTimeGetCurrent()
        var totalParticlesTransported: UInt64 = 0

        // Current fission source for the next batch
        var currentSource = source

        let activeBatches = numBatches - numInactive

        print("Starting k-eigenvalue simulation")
        print("  Particles per batch: \(numParticles)")
        print("  Inactive batches:    \(numInactive)")
        print("  Active batches:      \(activeBatches)")
        print("  Energy groups:       \(numGroups)")
        print("  Cells:               \(params.numCells)")
        print()

        batchKEff = []
        batchKEff.reserveCapacity(activeBatches)
        shannonEntropy = []

        for batchIdx in 0..<numBatches {
            let isActive = batchIdx >= numInactive

            // Reset per-batch state
            fissionBank.reset()
            tally.resetForBatch()
            memset(lostCountBuffer.contents(), 0, MemoryLayout<UInt32>.stride)

            // Initialize particles from current source
            seedFromSource(currentSource, batchIndex: UInt32(batchIdx))

            // Update params buffer with current k-eff
            writeParams()

            // Run transport loop until all particles are dead
            let iterations = transportBatch()

            // Read back fission bank
            let fissionSites = fissionBank.sites()
            let fissionCount = fissionSites.count

            // Compute batch k-eff: k_new = k_old * (fission_sites / num_particles)
            let batchK: Double
            if fissionCount > 0 {
                batchK = Double(params.kEff) * Double(fissionCount) / Double(numParticles)
            } else {
                batchK = Double(params.kEff)
            }

            // Collect tallies for active batches
            if isActive {
                tally.collectBatch()
                batchKEff.append(batchK)
            }

            // Shannon entropy for convergence diagnostics
            if !fissionSites.isEmpty {
                let entropy = KEigenvalue.shannonEntropy(
                    sites: fissionSites,
                    meshBins: (10, 10, 1),
                    bounds: (min: SIMD3<Float>(0, 0, 0),
                             max: SIMD3<Float>(1.26, 1.26, 1.0))
                )
                shannonEntropy.append(entropy)
            }

            // Update k-eff for next batch
            params.kEff = Float(batchK)

            // Prepare source for next batch from fission bank
            currentSource = resampleFissionBank(fissionSites)

            totalParticlesTransported += UInt64(numParticles)

            // After inactive phase, clear tally history
            if batchIdx == numInactive - 1 {
                tally.clearBatchHistory()
            }

            // Print progress
            let lostCount = lostCountBuffer.contents()
                .assumingMemoryBound(to: UInt32.self).pointee
            if verbose || batchIdx % 10 == 0 || batchIdx == numBatches - 1 {
                let status = isActive ? "Active  " : "Inactive"
                let convergedStr = KEigenvalue.isConverged(entropyHistory: shannonEntropy)
                    ? " [converged]" : ""
                print(String(format: "Batch %3d/%d [%@] k-eff = %.5f  fission = %d  lost = %d  iter = %d%@",
                             batchIdx + 1, numBatches, status, batchK,
                             fissionCount, lostCount, iterations, convergedStr))
            }
        }

        let wallTime = CFAbsoluteTimeGetCurrent() - wallStart
        let particlesPerSecond = Double(totalParticlesTransported) / wallTime

        // Compute statistics over active batches
        let kEffStats = Statistics.kEffStatistics(batchKEff: batchKEff)
        let fluxStats = Statistics.computeStats(batchValues: tally.batchFlux)
        let fissionStats = Statistics.computeStats(batchValues: tally.batchFission)

        return SimulationResult(
            kEff: kEffStats,
            fluxStats: fluxStats,
            fissionStats: fissionStats,
            wallTime: wallTime,
            particlesPerSecond: particlesPerSecond,
            totalParticles: totalParticlesTransported,
            numActiveBatches: activeBatches,
            gpuName: engine.device.name
        )
    }

    // MARK: - Private: Transport Loop

    /// Run a single batch's event-based transport loop until all particles are dead.
    /// Returns the number of iterations executed.
    private func transportBatch() -> Int {
        let maxSteps = 1000  // safety limit

        for step in 0..<maxSteps {
            let cmdBuffer = engine.commandQueue.makeCommandBuffer()!
            cmdBuffer.label = "Transport_\(step)"

            // 1. XS Lookup (particles with event == XS_LOOKUP -> DISTANCE)
            kernels.dispatchXSLookup(
                particles: particleBuffer,
                materials: materialsBuffer,
                cells:     geometry.cellBuffer,
                params:    paramsBuffer,
                count:     numParticles,
                commandBuffer: cmdBuffer
            )

            // 2. Distance to Collision (event == DISTANCE -> MOVE)
            kernels.dispatchDistanceToCollision(
                particles: particleBuffer,
                params:    paramsBuffer,
                count:     numParticles,
                commandBuffer: cmdBuffer
            )

            // 3. Move Particle (event == MOVE -> COLLIDE or XS_LOOKUP or DEAD)
            kernels.dispatchMove(
                particles:    particleBuffer,
                surfaces:     geometry.surfaceBuffer,
                cells:        geometry.cellBuffer,
                cellSurfaces: geometry.cellSurfaceBuffer,
                params:       paramsBuffer,
                lostCount:    lostCountBuffer,
                count:        numParticles,
                commandBuffer: cmdBuffer
            )

            // 4. Collision (event == COLLIDE -> TALLY)
            //    Buffer bindings match Collision.metal exactly:
            //    [0] particles, [1] materials, [2] fissionBank, [3] fissionCount, [4] params
            dispatchCollision(commandBuffer: cmdBuffer)

            // 5. Tally Score (event == TALLY -> XS_LOOKUP or DEAD)
            tally.dispatch(
                particles:     particleBuffer,
                params:        paramsBuffer,
                count:         numParticles,
                commandBuffer: cmdBuffer
            )

            cmdBuffer.commit()
            cmdBuffer.waitUntilCompleted()

            // Check if all particles are dead
            if allParticlesDead() { return step + 1 }
        }

        return maxSteps
    }

    /// Encode collision kernel into commandBuffer.
    ///
    /// Buffer bindings match Collision.metal:
    ///   [0] particles    (read-write)
    ///   [1] materials    (read-only, flat float array)
    ///   [2] fissionBank  (read-write, FissionSite array)
    ///   [3] fissionCount (read-write, atomic_uint counter)
    ///   [4] params       (read-only, SimParams)
    private func dispatchCollision(commandBuffer: MTLCommandBuffer) {
        let encoder = commandBuffer.makeComputeCommandEncoder()!
        encoder.setComputePipelineState(collisionPipeline)
        encoder.setBuffer(particleBuffer,          offset: 0, index: 0)
        encoder.setBuffer(materialsBuffer,         offset: 0, index: 1)
        encoder.setBuffer(fissionBank.buffer,      offset: 0, index: 2)
        encoder.setBuffer(fissionBank.countBuffer, offset: 0, index: 3)
        encoder.setBuffer(paramsBuffer,            offset: 0, index: 4)

        let gridSize = numParticles
        let maxThreads = collisionPipeline.maxTotalThreadsPerThreadgroup
        let threadgroupWidth = min(maxThreads, gridSize)
        encoder.dispatchThreads(
            MTLSize(width: gridSize, height: 1, depth: 1),
            threadsPerThreadgroup: MTLSize(width: threadgroupWidth, height: 1, depth: 1)
        )
        encoder.endEncoding()
    }

    /// Returns true when all particles have alive == 0.
    private func allParticlesDead() -> Bool {
        let ptr = particleBuffer.contents().bindMemory(
            to: Particle.self, capacity: numParticles)
        for i in 0..<numParticles {
            if ptr[i].alive != 0 { return false }
        }
        return true
    }

    // MARK: - Private: Source Seeding

    /// Initialize particles from a source distribution.
    ///
    /// Each particle gets:
    /// - Position from source bank (cycled if numParticles > source.count)
    /// - Isotropic direction via Philox RNG (reproducible per particle+batch)
    /// - Energy group from source
    /// - Initial cell from geometry test (C5G7 pincell: inside/outside fuel cylinder)
    private func seedFromSource(_ source: [(position: SIMD3<Float>, group: UInt32)],
                                batchIndex: UInt32) {
        let ptr = particleBuffer.contents().bindMemory(
            to: Particle.self, capacity: numParticles)

        for i in 0..<numParticles {
            let srcIdx = i % source.count
            var rng = PhiloxRNG(key: UInt32(i), counterHi: batchIndex)

            let dir = rng.sampleIsotropicDirection()

            var p = Particle(
                position: source[srcIdx].position,
                direction: dir
            )
            p.energyGroup      = source[srcIdx].group
            p.weight           = 1.0
            p.alive            = 1
            p.event            = ParticleEvent.xsLookup.rawValue
            p.rngCounter       = rng.counterLo
            p.rngKey           = UInt32(i) ^ (batchIndex &* 0x9E3779B9)
            p.boundarySurface  = -1
            p.distanceTraveled = 0
            p.fissionFlag      = 0

            // Determine initial cell from position.
            // C5G7 pincell: cell 0 = fuel (inside cylinder r=0.54 at center 0.63,0.63)
            //               cell 1 = moderator (outside cylinder)
            let cx: Float = 0.63
            let cy: Float = 0.63
            let r: Float = 0.54
            let dx = p.position.x - cx
            let dy = p.position.y - cy
            p.cellIndex = (dx * dx + dy * dy < r * r) ? 0 : 1

            ptr[i] = p
        }
    }

    // MARK: - Private: Fission Bank Resampling

    /// Resample fission bank sites to create a new source for the next batch.
    /// Falls back to the default C5G7 source sampler if the bank is empty.
    private func resampleFissionBank(
        _ sites: [FissionSite]
    ) -> [(position: SIMD3<Float>, group: UInt32)] {
        guard !sites.isEmpty else {
            return C5G7Geometry.sourceSampler(
                numParticles: numParticles, numGroups: numGroups)
        }

        var source: [(position: SIMD3<Float>, group: UInt32)] = []
        source.reserveCapacity(numParticles)

        for i in 0..<numParticles {
            let idx = i % sites.count
            source.append((position: sites[idx].position,
                           group: sites[idx].energyGroup))
        }
        return source
    }

    // MARK: - Private: Parameter Update

    /// Write current params struct to the GPU buffer.
    private func writeParams() {
        let ptr = paramsBuffer.contents()
            .assumingMemoryBound(to: SimulationParams.self)
        ptr.pointee = params
    }
}
